### PR TITLE
Grammar change - Allow AliasDeclarationY to express function types

### PR DESCRIPTION
**Pull request by ghost**
fix issue 16020 - Allow AliasDeclarationY to express function types

The (not so) new alias declaration, using the assignment operator, is preferred over the old C-style declarations.
However it didn't allow to express function types, which lead to blend the declaration styles.

This change adds a new rule to `AliasDeclarationY` to solve the problem,

```diff
AliasDeclarationY:
+    'alias' Identifier '=' StorageClasses? BasicType '(' Parameters ')' Attributes?
```

consolidating the language consistency.

### CODE REVIEW

File: src/dmd/parse.d
Comment (jacob-carlborg) [inline src/dmd/parse.d]: What about other than basic types?

Comment (Basile-z) [reply src/dmd/parse.d]: This is a pragmatic choice. There's no lookup function for `Type`.

Comment (jacob-carlborg) [reply src/dmd/parse.d]: is this handled the same way in other places in the parser?

Comment (Basile-z) [reply src/dmd/parse.d]: I don't know but I'll will have to do it anyway since the C-style syntax support a full Type (note that the only difference is the random amount of TypeCtors before the BasicType).

**Comment by rainers** [comment]
A function can be specified by
```
alias fun = void function();
```
The syntax without `function` is a bit ambguous as it could also be a delegate.

**Comment by Basile-z** [comment]
> A function can be specified by

No what you declare is a function **pointer** type not a function type.

File: src/dmd/parse.d
Comment (Basile-z) [reply src/dmd/parse.d]: Actually the type ctors are part of the function type, not the return. So it's okay to just parse a BasicType. Then the type ctors can be placed as postfixes.

**Comment by rainers** [comment]
> No what you declare is a function **pointer** type not a function type, i.e the type of `&stuff` not the type of `stuff`.

Ah, ok, I misunderstood. Should have read the bug report before commenting. Being able to specify the type directly sounds good. Even already supported by dparser/Visual D ;-)

**Comment by AndrejMitrovic** [comment]
I remember this part being missing from the language. I don't recall what I needed it for back in the days, maybe for interfacing with C. 👍

What are some use-cases? Tther than just declaring aliases, how/where are these actually used?

File: src/dmd/parse.d
Comment (JinShil) [inline src/dmd/parse.d]: I think this should be `isTypeCtor` without the trailing 's'.

File: src/dmd/parse.d
Comment (JinShil) [inline src/dmd/parse.d]: I prefer `while(true)`

**Review by JinShil — APPROVED** [review]

File: src/dmd/parse.d
Comment (JinShil) [reply src/dmd/parse.d]: Is there some way you can consolidate this function and https://github.com/dlang/dmd/blob/55ca5920ca0f7da1f4096eafb95feaffa6813a79/src/dmd/parse.d#L1518-L1552 to remove the reduandancy an duplication?

**Review by JinShil — CHANGES_REQUESTED** [review]

File: src/dmd/parse.d
Comment (Basile-z) [reply src/dmd/parse.d]: I put a s because it's for a pack of TypeCtor. Actually if you have just one an any bad token following the func returns false. It's really to detect the pack that can be found before a BasicType.

Comment (Basile-z) [reply src/dmd/parse.d]: I'm aware that it can look redundant but the problem is that parseTypeCtor works on the current token while isTypeCtor is a lookup function.

Comment (JinShil) [reply src/dmd/parse.d]: What about renaming `isTypeCtor()` something like `advanceToTypeCtor` and moving the `case` statement to a ` bool isTypeCtor(TOK value)`?  Then `advanceToTypeCtor` and `parseTypeCtor` can both call `isTypeCtor`.  Or something like that.

Comment (JinShil) [reply src/dmd/parse.d]: Better yet, we should be able to pull out the whole `while` loop into it's own function.  I can't work out the details right now, but D's such a powerful language.  There's got to be a way.

Comment (Basile-z) [reply src/dmd/parse.d]: I can simplify the lookup. Just look at next token. If it's a type ctor that's not followed by a left paren then call parseTypeCtor. This will work.

Comment (Basile-z) [reply src/dmd/parse.d]: No that didn't work

**Comment by Basile-z** [comment]
> What are some use-cases? Other than just declaring aliases, how/where are these actually used?

It can be used to match Aggregate member with function types, independently from the context.
There are several use case of the current syntax in std, for example https://github.com/dlang/phobos/blob/c6bfb9fe19ee3bb4bdc63bcb5fb75837d905e414/std/traits.d#L2301

**Comment by jacob-carlborg** [comment]
> What are some use-cases? Other than just declaring aliases, how/where are these actually used?

I had a use case just a couple of days ago. I have a template take takes to functions by `alias` parameters. I want to make sure the signatures are exactly the same except that the second function should have one additional parameter. Since the types of the functions are function types and not function pointers I need a way declare a function type to be able to verify the signatures are correct. Now I have a workaround where I take the address of one of the `alias` parameters to get a function pointer instead of a function.

https://github.com/jacob-carlborg/dlp/blob/0c4901ab8dd0b4af500be6becbf3239b0f1130c5/source/dlp/core/redirect.d

**Comment by Geod24** [comment]
> I remember this part being missing from the language. I don't recall what I needed it for back in the days, maybe for interfacing with C. 👍

I remember we had this problem in one of the library while converting from D1 to D2.
IIRC we ended up not really needing it anyway. But there is an easy way to express this:

```d
alias Fun = typeof(*(void function()).init);
pragma(msg, Fun);
```

I don't think we need an extra, possibly ambiguous syntax for this.

**Comment by Basile-z** [comment]
> I don't think we need an extra, possibly ambiguous syntax for this.

I think that many seems to misunderstand what proposes this PR. The feature is there and working, already. But the syntax is not conform to the style that most people use nowadays. This is just a syntactic sugar, the semantic remains untouched.

File: src/dmd/parse.d
Comment (JinShil) [inline src/dmd/parse.d]: I'm sorry, but I don't consider the matter of consolidating this code with https://github.com/dlang/dmd/blob/55ca5920ca0f7da1f4096eafb95feaffa6813a79/src/dmd/parse.d#L1518-L1552 resolved.  At a minimum, the `case` statement can be put into a function shared by both implementations (perhaps somewhere near the `TOK` enum declaration).

**Review by JinShil — APPROVED** [review]

File: src/dmd/parse.d
Comment (JinShil) [reply src/dmd/parse.d]: Dammit.  Hit the wrong button.  I mean't to select "Request Changes"

**Comment by Basile-z** [comment]
@WalterBright @andralex, can this be formally approved or rejected now ? I'd like to have your pov on this possible change now that everything is well explained.

File: src/dmd/parse.d
Comment (WalterBright) [inline src/dmd/parse.d]: I have no idea what a "pack" is. Need to:

1. give an example of what a "pack" is
2. give URL to spec on what is going on here

Comment (WalterBright) [reply src/dmd/parse.d]: Also rephrase the comment to use the Ddoc `returns: ... explanation...` style.

Comment (Basile-z) [reply src/dmd/parse.d]: Okay. By pack I meant a "sequence of".

Comment (Basile-z) [reply src/dmd/parse.d]: function removed, as suggested by Mike F.

File: src/dmd/parse.d
Comment (Basile-z) [reply src/dmd/parse.d]: I've managed to apply your suggestion so this is resolved for good this time

**Comment by wilzbach** [comment]
Needs a changelog entry and update PR to dlang.org

**Comment by Basile-z** [comment]
@wilzbach done, you can remove the tags.

File: changelog/aliasdecly_func.dd
Comment (ZombineDev) [inline changelog/aliasdecly_func.dd]:
```suggestion
The declaration consists of return type, followed by parameters and ends with optional (member) function attributes.
```

**Review by PetarKirov — APPROVED** [review]
Nice work!

**Comment by Basile-z** [comment]
You could remove the other tag, the spec PR is there : https://github.com/dlang/dlang.org/pull/2632

File: src/dmd/parse.d
Comment (wilzbach) [inline src/dmd/parse.d]: I think Walter doesn't like reusing variables and prefers declaring new ones.

Comment (WalterBright) [reply src/dmd/parse.d]: that's right.

File: changelog/aliasdecly_func.dd
Comment (WalterBright) [inline changelog/aliasdecly_func.dd]: 1. What is `AliasDeclarationY` ?
2. Need to precisely express grammar changes using BNF, like the spec does, not handwaving.

File: src/dmd/parse.d
Comment (WalterBright) [inline src/dmd/parse.d]: No documentation on what the following grammar is being parsed.

**Review by WalterBright — DISMISSED** [review]
Need documentation per my comments.

File: changelog/aliasdecly_func.dd
Comment (WalterBright) [reply changelog/aliasdecly_func.dd]: Also the statement makes no grammatical sense.

**Comment by WalterBright** [comment]
Also, since this is a significant language change, shouldn't it come with at DIP justifying it?

**Comment by WalterBright** [comment]
Language changes should **never** be "72h no object -> merge".

**Comment by JinShil** [comment]
> Language changes should never be "72h no object -> merge".

In this case, I think the "72h no object -> merge" tag worked exactly as it was designed to work.  We lack a formal procedure to ensure **ALL** PRs receive a **timely** review.  Until that is implemented, "72h no object -> merge" is the best we have.  Without it, PRs will just be ignored indefinitely.

**Comment by JinShil** [comment]
> Also, since this is a significant language change, shouldn't it come with at DIP justifying it?

In general, yes, but the DIP process is heavy in process and formalities.  It consumes a significant amount of effort from the author, the DIP manager, the community, and the language authors.  This change is quite isolated and simple compared to other, more significant, language changes.  So, in this case, the DIP process seems disproportionate.

That being said, in this case, I do sense a danger of introducing unintended consequences if all corner cases are not accounted for.  I can't identify any, but that doesn't mean there isn't one.  Having a broader peer review would be nice.

I hope at the upcoming DConf, there will be some discussion about having a more proportionate approval process for changes like this.  cc @mdparker

**Comment by thewilsonator** [comment]
Yep a stratified DIP process is on the list of things to do.

**Comment by WalterBright** [comment]
> Without it, PRs will just be ignored indefinitely.

I didn't even know about this PR till a couple hours ago. It didn't show up in the n.g. There's been no discussion. Mike Parker is in charge of this stuff and he was not notified.

This PR has little justification presented, no alternatives presented, no precision (i.e. no grammar written for the new grammar),  etc. You're basically dumping the work onto me. That's not how work is saved.

Doing "72 hours" for this kind of thing is nonsense. It's likely ok for minor things. Not language changes.

**Comment by JinShil** [comment]
> I didn't even know about this PR till a couple hours ago. 

Yes, that's a serious problem.  Why were you not aware of it?  It's not 72h old; it's 6 days old.  I check the [DMD PR queue](https://github.com/dlang/dmd/pulls) daily.  I also receive notifications for **ALL** DMD PR activity because I am "Watch"ing the DMD repository.

image

No DMD PR should go unnoticed by those who make it a priority to pay attention and utilize GitHub's notification features.

**Comment by WalterBright** [comment]
> I do sense a danger of introducing unintended consequences if all corner cases are not accounted for. I can't identify any, but that doesn't mean there isn't one.

That's a very good reason why language changes should not be rushed with 72 hr deadlines and no notice to people who have experience with this sort of stuff. Please keep in mind that if it turns out to have unfortunate consequences, we'd be stuck with it for YEARS.

Note that I haven't made any determination if this idea is a good one or not. I'm complaining about the process here.

**Comment by WalterBright** [comment]
> No DMD PR should go unnoticed by those who make it a priority to pay attention and utilize GitHub's notification features.

I don't expect to have significant language features pushed through without notification to the review manager, Mike Parker. The subject of the PR does not indicate a language change, either. (Yes, it's tagged as an enhancement, but that could mean a lot of things.)

**Comment by jacob-carlborg** [comment]
> I didn't even know about this PR till a couple hours ago

@WalterBright  you did. You commented on this PR four days ago: https://github.com/dlang/dmd/pull/9667#pullrequestreview-229789250.

> Also, since this is a significant language change, shouldn't it come with at DIP justifying it?

This is not a significant language change. It's a minor addition. This syntax is already possible today:

```d
alias int foo(int);
```

The  change in this PR allows to use the newer alias syntax, `alias =`, as in:

```d
alias foo = int(int);
```

**Comment by wilzbach** [comment]
> This is not a significant language change. It's a minor addition. This syntax is already possible today:

Something that has been blocking the deprecation of the old alias syntax for years and was definitely overlooked when `alias` was introduced, hence small bug fix.

**Review by andralex — COMMENTED** [review]
This is a good change once @WalterBright 's points have been addressed. Thanks!

**Comment by WalterBright** [comment]
> @WalterBright you did. You commented on this PR four days ago

So I did. You're right.

File: src/dmd/parse.d
Comment (Basile-z) [reply src/dmd/parse.d]: It's not reused. It's assigned only once in each branch handling a particular form of alias.

File: src/dmd/parse.d
Comment (Basile-z) [reply src/dmd/parse.d]: fixed

**Comment by Basile-z** [comment]
@WalterBright, waiting for a dismiss of your change request (i.e approval) or more changes to apply.

**Comment by n8sh** [comment]
>The (not so) new alias declaration, using the assignment operator, is preferred over the old C-style declarations.
>However it didn't allow to express function types

What do you mean? This works fine with the current alias syntax.
```d
alias CurrentStyleFunAlias = int function(string);
```

**Comment by Basile-z** [comment]
n8sh, what you declare is a function **pointer** type not a function type. Read the previous comments, you're not the first to confuse both.

**Comment by n8sh** [comment]
I see, thanks.

**Comment by thewilsonator** [comment]
@WalterBright readding 72h->merge

File: changelog/aliasdecly_func.dd
Comment (Basile-z) [inline changelog/aliasdecly_func.dd]: wait before merge, the links won't work

Comment (Basile-z) [reply changelog/aliasdecly_func.dd]: fixed

**Comment by WalterBright** [comment]
The current `AliasDeclarationY` is https://dlang.org/spec/function.html#MemberFunctionAttributes :
```
AliasDeclarationY:
    Identifier TemplateParametersopt = StorageClassesopt Type
    Identifier TemplateParametersopt = FunctionLiteral
    Identifier TemplateParametersopt = StorageClassesopt BasicType Parameters MemberFunctionAttributesopt
```


You propose adding:
```
   'alias' Identifier '=' StorageClasses? BasicType '(' Parameters ')' Attributes?
```
The first thing is that `Parameters` https://dlang.org/spec/function.html#Parameters already include parentheses. The second thing is I don't see what the new grammar rule is adding. Looks to me like it already supports function types. Also, new grammar rules proposed should fit into the existing use of names, not handwave - use MemberFunctionAttributes, not Attributes, and don't omit TemplateParameters.

**Review by WalterBright — DISMISSED** [review]
Proposed new grammar rule doesn't clarify what new grammar is being introduced and is not consistent with existing grammar rules.

**Comment by Basile-z** [comment]
alias equal something else is better than what ever exists now. The syntax is clear, you have something assigned to something else. This is not obvious with the old syntax.

**Comment by rainers** [comment]
```
> AliasDeclarationY:
>     ...
>     ...
>     Identifier TemplateParametersopt = StorageClassesopt BasicType Parameters MemberFunctionAttributesopt
```
@WalterBright To clarify: this third rule has been merged into the documentation (a bit prematurely) as a companion to this PR. The initial description is outdated.

@Basile-z The grammar for the old syntax allowed templates parameters in the declarator (not sure how that worked). Does this implementation allow it too or is the omission in the grammar deliberate?

**Comment by Basile-z** [comment]
The fear of change will bring you into darkness

**Comment by WalterBright** [comment]
> @WalterBright To clarify: this third rule has been merged into the documentation (a bit prematurely) as a companion to this PR. The initial description is outdated.

Thank you. Neither the bugzilla issue nor this PR has a link to that. Here it is: https://github.com/dlang/dlang.org/pull/2632

In general, language changes need to be approved by myself.

**Review by WalterBright — APPROVED** [review]

**Comment by Basile-z** [comment]
What's happening  ? I had the feeling that this was about to be rejected, hence my tragedy-grade comment.

> Proposed new grammar rule doesn't clarify what new grammar is being introduced and is not consistent with existing grammar rules.

Honestly I have no idea how to address this. I've searched for previous case of grammar rules added in the changelog. There's none. I'd use the "diff" coloring if available.

**Pull request merged**
